samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -274,7 +276,8 @@ public void onContainerFailed(Throwable t) {
               } else {
                 LOGGER.warn("JobCoordinatorLatch was null. It is possible for some component to be waiting.");
               }
-              LOGGER.error("Container failed. Stopping the processor.", t);
+              containerException = t;
+              LOGGER.error("Container failed. Stopping the processor.", containerException);
               container = null;
               stop();
```

Comment:
>Another problem, that this will cause the stop() to run in the a different thread (container thread).

Yes. We knew that it will be invoked from a different thread. We assumed that all tasks in the ZkJobCoordinator queue would stop when the executor service is shutdown. However, there isn't a lot we can do other than interrupting the thread. 

> May be we should invoke the stop from JobCoordinator on a debounce Thread?

Yes. That is an option we can explore. Thanks for creating jiras to track this issue.